vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,787 @@
+#include <memory>
+#include <sstream>
+#include <type_traits>
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+
+namespace VW
+{
+namespace reductions
+{
+namespace eigen_memory_tree
+{
+////////////////////////////definitions/////////////////////
+////////////////////////////////////////////////////////////
+emt_example::emt_example() { label = 0; }
+
+emt_example::emt_example(VW::workspace& all, VW::example* ex)
+{
+  label = ex->l.multi.label;
+
+  std::vector<std::vector<VW::namespace_index>>* full_interactions = ex->interactions;
+  std::vector<std::vector<VW::namespace_index>> base_interactions;
+
+  ex->interactions = &base_interactions;
+  auto* ex1 = VW::flatten_sort_example(all, ex);
+  for (auto& f : ex1->fs) { base.emplace_back(f.index(), f.value()); }
+  VW::free_flatten_example(ex1);
+
+  ex->interactions = full_interactions;
+  auto* ex2 = VW::flatten_sort_example(all, ex);
+  for (auto& f : ex2->fs) { full.emplace_back(f.index(), f.value()); }
+  VW::free_flatten_example(ex2);
+}
+
+emt_lru::emt_lru(unsigned long max_size) { (*this).max_size = max_size; }
+
+emt_lru::K emt_lru::bound(emt_lru::K item)
+{
+  if (max_size == 0) { return nullptr; }
+
+  auto item_map_reference = map.find(item);
+
+  // item is not in map so we add it to our list
+  if (item_map_reference == map.end())
+  {
+    list.push_front(item);
+    map.emplace(std::pair<K, V>(item, list.begin()));
+  }
+  // item is in map already so we move it to the front of the line
+  else
+  {
+    V item_list_reference = (*item_map_reference).second;
+    if (item_list_reference != list.begin())
+    {
+      list.splice(list.begin(), list, item_list_reference, std::next(item_list_reference));
+    }
+  }
+
+  if (list.size() > max_size)
+  {
+    K last_value = list.back();
+    list.pop_back();
+    map.erase(last_value);
+    return last_value;
+  }
+  else { return nullptr; }
+}
+
+emt_node::emt_node()
+{
+  left = nullptr;
+  right = nullptr;
+  router_decision = 0;
+}
+
+emt_tree::emt_tree()
+{
+  leaf_split = 100;
+
+  scorer_type = emt_scorer_type::self_consistent_rank;
+  router_type = emt_router_type::eigen;
+
+  begin = 0;
+
+  all = nullptr;
+  ex = nullptr;
+  bounder = nullptr;
+  root = VW::make_unique<emt_node>();
+
+  // we set this up for repeated use later in the scorer.
+  // we will populate this examples features over and over.
+  ex = VW::make_unique<VW::example>();
+  ex->interactions = new std::vector<std::vector<VW::namespace_index>>();
+  ex->extent_interactions = new std::vector<std::vector<extent_term>>();
+  ex->indices.push_back(0);
+}
+
+emt_tree::~emt_tree()
+{
+  delete ex->interactions;
+  delete ex->extent_interactions;
+}
+
+////////////////////////end of definitions/////////////////
+///////////////////////////////////////////////////////////
+
+///////////////////////Helper/////////////////////////////
+//////////////////////////////////////////////////////////
+float emt_median(std::vector<float>& array)
```

Comment:
This can be a const input

Reply:
@jackgerrits Can it? My compiler complains. I think it is because we re-order the array?

Reply:
You're right, I didn't realize `std::nth_element` mutates the input.